### PR TITLE
Assignment6 @Malcolm-Shepherd

### PR DESCRIPTION
**Pull request by Clockknight**

### CODE REVIEW

**Comment by Clockknight** [comment]
Still working on the program, but getting in PR before deadline

**Comment by FenrisStargazer** [comment]
I will review this

**Review by FenrisStargazer — COMMENTED** [review]
**Instructions**
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ❌?✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ❌?✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ❌
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ❌
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ❌

**Extra Credit**
Refactor the redirect portion of the Program class into 'ProgramBase`❌
Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects❌
Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)❌

**Fundamentals**
Place all shared project properties into a Directory.Build.Props file.✔
Place all shared project items into a Directory.Build.targets file.❌
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔ _(kind of)_
Choose simplicity over complexity ✔

I was unable to pull down your code to try running it, but at a glance I don't see anything that would cause a catastrophic error, though there may be warnings as the IDE is nitpicky, I'm not sure. Likewise, at a glance, all your tests I believe will pass without trouble, if perhaps only by technicality. You've a very strong start here but there is room for some refinement and adjustments.

File: ClassLibraryProjects/Calculate/Calculator.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            operators.Add('+', Add);
+            operators.Add('-', Subtract);
+            operators.Add('*', Multiply);
+            operators.Add('/', Divide);
+
+            return operators;
+        } 
+    }
+
+    public int? TryCalculate(string calculation)
```
[the remark this answers is not in the document]
Comment (Clockknight) [reply ClassLibraryProjects/Calculate/Calculator.cs:30]: Ah, I see. I'll be sure to try that - I don't think I understood what it meant by "following the TryParse pattern".

**Comment by JessicaH46** [comment]
I will review this

File: ClassLibraryProjects/Calculate/Program.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+public class Program
+{
+    public Action<string> WriteLine{ 
+        get 
+        { 
+            Action<string> action = Console.WriteLine;
+            return action;
+        }
+        init
```
Comment (JessicaH46) [inline ClassLibraryProjects/Calculate/Program.cs:14]: I think another way you can do it that might be a little more simplified
`public Action<string> WriteLine { get; init; } = Console.WriteLine;
  public Func<string> ReadLine { get; init; } = Console.ReadLine;`

**Comment by JessicaH46** [comment]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ❌✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ❌✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ❌
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ❌
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ❌✔

I think your code looks really good so far! A little confused on the 2nd Program.cs file that you made, but I think it's just. testing file you forgot to remove. I think you might want to change its name if you want to leave it in there so there isn't any confusion on the file names. I think everything you have gotten done so far looks goo and is heading in the right direction. You already have most of the main components done so that is good! Good work!

File: ClassLibraryProjects/ConsoleApp1/Program.cs [hunk 3]
```diff
@@ -0,0 +1,2 @@
+﻿// See https://aka.ms/new-console-template for more information
+Console.WriteLine("Hello, World!");
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/ConsoleApp1/Program.cs:2]: I don't know why this is here. I think it was deleted from the .sln but not actually removed.

File: ClassLibraryProjects/CalculateTests/CalculatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace CalculateTests;
+
+[TestClass]
+public class CalculatorTests
+{
+    [TestMethod]
+    public void OnConstruction_CalculatorProperties_Pass()
+    {
+        Calculator testCalculator = new();
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/CalculateTests/CalculatorTests.cs:13]: Test deleted, a later test was supposed to replace this one.

File: ClassLibraryProjects/CalculateTests/CalculatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    [TestMethod]
+    public void ProgramWriteLine_WritesToConsole_Pass()
+    {
+        string testLine = "I'm on the console.";/*
+        Action<string> write = (string text) => { Console.WriteLine(text); };*/
+        Program testPr = new();
+        testPr.WriteLine(testLine);
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/CalculateTests/CalculatorTests.cs:22]: I thought Mark said at some point in class that we didn't have to test things like Console.WriteLine, so I didn't full finish this test. Fixed this and the ReadLine test.

File: ClassLibraryProjects/CalculateTests/CalculatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        Assert.IsTrue(testCalculator.MathematicalOperations.TryGetValue('-', out testSub!));
+        Assert.IsTrue(testCalculator.MathematicalOperations.TryGetValue('*', out testMult!));
+        Assert.IsTrue(testCalculator.MathematicalOperations.TryGetValue('/', out testDiv!));
+
+
+    }
+
+
+    [TestMethod]
+    public void TryCalculate_ParseCommands_Pass()
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/CalculateTests/CalculatorTests.cs:53]: I agree with both of these points, refactored into 2 tests and changed this test's expected to an expected value.

File: ClassLibraryProjects/Calculate/Program.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        }
+    }
+
+    public Program() { }
+
+    public static void Main(string[] args) 
+    {
+        Calculator calculator = new Calculator();
+        Program program = new Program();    
+        int? result = 0;
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/Calculate/Program.cs:39]: Fixed to use a bool return and an out int parameter

File: ClassLibraryProjects/Calculate/Program.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+        while(result is not null) 
+        {
+            program.WriteLine("Enter an equation:");
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/Calculate/Program.cs:43]: It was designed with the idea that the user knew how to input for this assignment, but if it needed to be more robust I would agree.

File: ClassLibraryProjects/Calculate/Calculator.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            Dictionary<char, Func<int, int, int>> operators = new();
+            operators.Add('+', Add);
+            operators.Add('-', Subtract);
+            operators.Add('*', Multiply);
+            operators.Add('/', Divide);
+
+            return operators;
+        } 
+    }
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/Calculate/Calculator.cs:28]: We have a test that shows this does work as intended. It also doesn't add duplicates as a new Dictionary is created and added to. We aren't very familiar with Dictionaries however, so there is probably a better way to do this.

File: ClassLibraryProjects/Calculate/Calculator.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+                return Calculator.Add(operand1, operand2);
+            case "-":
+                return Calculator.Subtract(operand1, operand2);
+            case "*":
+                return Calculator.Multiply(operand1, operand2);
+            case "/":
+                return Calculator.Divide(operand1, operand2);
+            default:
+                return null;
+        }
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/Calculate/Calculator.cs:62]: Rewrote to use MathematicalOperations instead

File: ClassLibraryProjects/Calculate/Program.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+using System.Runtime.CompilerServices;
+
+public class Program
+{
+    public Action<string> WriteLine{ 
+        get 
+        { 
+            Action<string> action = Console.WriteLine;
+            return action;
+        }
```
[the remark this answers is not in the document]
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/Calculate/Program.cs:13]: Fixed using solution from another comment

File: ClassLibraryProjects/Calculate/Program.cs (hunk 2 above)
Comment (Malcolm-Shepherd) [reply ClassLibraryProjects/Calculate/Program.cs:14]: Fixed, this is definitely simpler and more readable.

File: ClassLibraryProjects/Calculate/Program.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+using Calculate;
+using System.Runtime.CompilerServices;
+
+public class Program
+{
+    public Action<string> WriteLine{ get; init; } = Console.WriteLine;
+
+    public Func<string> ReadLine{ get; init; } = Console.ReadLine!;
+
+    public Program() { }
```
Comment (mmwoodfo) [inline ClassLibraryProjects/Calculate/Program.cs:12]: You should also have a constructor to set the write/read line properties

File: ClassLibraryProjects/Calculate/Calculator.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (mmwoodfo) [reply ClassLibraryProjects/Calculate/Calculator.cs:28]: It is valid, but... it's also something that could be done in a constructor. Reference @FenrisStargazer's PR to see how that code works, as it uses a constructor.

File: ClassLibraryProjects/Calculate/Calculator.cs [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            operators.Add('/', Divide);
+
+            return operators;
+        } 
+    }
+
+    public bool TryCalculate(string calculation, out int result)
+    {
+        result = 0;
+        string[] calcvars = calculation.Split(' ');
```
Comment (mmwoodfo) [inline ClassLibraryProjects/Calculate/Calculator.cs:33]: You should be using the Split method along side pattern matching. e.g.
`if (expression?.Split(" ") is [string operator1, [char mathmaticalOperator], string operand2])`

File: ClassLibraryProjects/CalculateTests/CalculatorTests.cs [hunk 14]
```diff
@@ -0,0 +1,99 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+using Calculate;
+
+namespace CalculateTests;
+
+[TestClass]
```
Comment (mmwoodfo) [inline ClassLibraryProjects/CalculateTests/CalculatorTests.cs:7]: could use some tests for dividing by 0, this is also a great place for data driven tests. Refer to @VeraKorchemnaya and  @AndrewKorchemniy's  PR for a good example of this.

**Pull request closed without merging**